lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2262,6 +2262,174 @@ class ArchiveTool : public Tool {
+      }
+      return true;
+    }
+
+    ctx.error("unexpected attribute: '" + name + "'");
+    return false;
+  }
+  virtual bool configureAttribute(const ConfigureContext& ctx, StringRef name,
+                                  ArrayRef<std::pair<StringRef, StringRef>> values) override {
+    // No supported attributes.
+    ctx.error("unexpected attribute: '" + name + "'");
+    return false;
+  }
+
+  virtual BuildValue getResultForOutput(Node* node,
+                                        const BuildValue& value) override {
+    // If the value was a failed command, propagate the failure.
+    if (value.isFailedCommand() || value.isPropagatedFailureCommand() ||
+        value.isCancelledCommand())
+      return BuildValue::makeFailedInput();
+    if (value.isSkippedCommand())
+      return BuildValue::makeSkippedCommand();
+
+    // Otherwise, this was successful, return the value as-is.
+    return BuildValue::fromData(value.toData());;
+  }
+
+  virtual bool isResultValid(BuildSystem& system,
+                             const BuildValue& value) override {
+    // Always re-run stale file removal.
+    return false;
+  }
+
+  virtual void start(BuildSystemCommandInterface& bsci,
+                     core::Task* task) override {}
+
+  virtual void providePriorValue(BuildSystemCommandInterface&, core::Task*,
+                                 const BuildValue& value) override {
+    hasPriorResult = true;
+    priorValue = BuildValue::fromData(value.toData());
+  }
+
+  virtual void provideValue(BuildSystemCommandInterface&, core::Task*,
+                            uintptr_t inputID,
+                            const BuildValue& value) override {
+    assert(0 && "unexpected API call");
+  }
+
+  virtual BuildValue execute(BuildSystemCommandInterface& bsci,
+                             core::Task* task,
+                             QueueJobContext* context) override {
+    // Nothing to do if we do not have a prior result.
+    if (!hasPriorResult) {
+      bsci.getDelegate().commandStarted(this);
+      bsci.getDelegate().commandFinished(this);
+      return BuildValue::makeStaleFileRemoval(expectedOutputs);
+    }
+
+    if (!priorValue.isStaleFileRemoval()) {
+      return BuildValue::makeFailedCommand();
+    }
+
+    std::vector<StringRef> priorNodes = priorValue.getDirectoryContents();
+
+    std::sort(priorNodes.begin(), priorNodes.end());
+    std::sort(expectedOutputs.begin(), expectedOutputs.end());
+
+    std::set_difference(priorNodes.begin(), priorNodes.end(),
+                        expectedOutputs.begin(), expectedOutputs.end(),
+                        std::back_inserter(filesToDelete));
+
+    bsci.getDelegate().commandStarted(this);
+
+    for (auto fileToDelete : filesToDelete) {
+      if (!getBuildSystem(bsci.getBuildEngine()).getDelegate().getFileSystem().remove(fileToDelete)) {
+        bsci.getDelegate().error("", {}, (Twine("cannot remove stale file '") + fileToDelete + "': " + strerror(errno)));
+      }
```

Comment:
Need to do this in a follow-up, because right now the delegate doesn't support arbitrary log output, but only output from a command process.